src/uvc_ros_driver.cpp

```diff
@@ -40,1372 +40,899 @@
+    ROS_ERROR("Unable to open the device");
+    return res;
+  }
+
+  /* Try to negotiate a 640x480 30 fps YUYV stream profile */
+  res = uvc_get_stream_ctrl_format_size(
+      devh_, &ctrl_,              /* result stored in ctrl */
+      UVC_FRAME_FORMAT_YUYV,      /* YUV 422, aka YUV 4:2:2. try _COMPRESSED */
+      raw_width_, raw_height_, 30 /* width, height, fps */
+      );
+
+  if (res < 0) {
+    /* device doesn't provide a matching stream */
+    uvc_perror(res, "get_mode");
+    ROS_ERROR("Device doesn't provide a matching stream");
+    return res;
+  }
+
+  return res;
 }
 
-////////////////////////////////////////////////////////////////////////////////
+uint8_t *uvcROSDriver::rawDataPtr(const uvc_frame_t *frame, const size_t line,
+                                  const size_t offset) {
+  return &(static_cast<uint8_t *>(
+      frame->data)[2 * ((line + 1) * frame->width - offset)]);
+}
 
-int16_t uvcROSDriver::ShortSwap(int16_t s)
-{
-	unsigned char b1, b2;
-	b1 = s & 255;
-	b2 = (s >> 8) & 255;
-	return (b1 << 8) + b2;
+uint8_t uvcROSDriver::readUInt8(const uvc_frame_t *frame, const size_t line,
+                                const size_t offset) {
+  return *rawDataPtr(frame, line, offset);
 }
 
-////////////////////////////////////////////////////////////////////////////////
+int16_t uvcROSDriver::readInt16(const uvc_frame_t *frame, const size_t line,
+                                const size_t offset) {
+  const uint8_t *data = rawDataPtr(frame, line, offset);
+  uint16_t value = static_cast<uint16_t>((data[1] << 0) | (data[0] << 8));
+  return *reinterpret_cast<int16_t *>(&value);
+}
+
+uint32_t uvcROSDriver::readUInt32(const uvc_frame_t *frame, const size_t line,
+                                  const size_t offset) {
+  const uint8_t *data = rawDataPtr(frame, line, offset);
 
-inline void uvcROSDriver::selectCameraInfo(int camera,
-		sensor_msgs::CameraInfo **ci)
-{
-	switch (camera) {
-	case 0:
-		*ci = &info_cam_0_;
-		break;
-
-	case 1:
-		*ci = &info_cam_1_;
-		break;
-
-	case 2:
-		*ci = &info_cam_2_;
-		break;
-
-	case 3:
-		*ci = &info_cam_3_;
-		break;
-
-	case 4:
-		*ci = &info_cam_4_;
-		break;
-
-	case 5:
-		*ci = &info_cam_5_;
-		break;
-
-	case 6:
-		*ci = &info_cam_6_;
-		break;
-
-	case 7:
-		*ci = &info_cam_7_;
-		break;
-
-	case 8:
-		*ci = &info_cam_8_;
-		break;
-
-	case 9:
-		*ci = &info_cam_9_;
-		break;
-	}
+  const uint32_t value =
+      (data[3] << 0) | (data[2] << 8) | (data[1] << 16) | (data[0] << 24);
+  return value;
 }
 
-////////////////////////////////////////////////////////////////////////////////
+bool uvcROSDriver::extractAndTranslateTimestamp(size_t line,
+                                                bool update_translator,
+                                                uvc_frame_t *frame,
+                                                ros::Time *stamp) {
+  // read out micro second timestamp
+  constexpr size_t kTimestampOffset = 10;
```

Comment:
why do you have this consexpr here in the function and not maybe in the header file?

Reply:
My idea behind having the constexprs being local to each function was because they are generally very specific to the operation being performed. I thought that it would be cleaner to define them next to where they are used rather than have a block of 30 constants at the top. I can see arguments for moving them out though.